src/core/cells.cpp

```diff
@@ -79,29 +82,66 @@ std::vector<std::pair<int, int>> get_pairs(double distance) {
   return ret;
 }
 
-void mpi_get_pairs_local(int, int) {
-  on_observable_calc();
+std::vector<std::pair<int, int>> get_pairs(double const distance) {
+  return get_pairs_filtered(distance, [](Particle const &p) { return true; });
+}
 
-  double distance;
-  boost::mpi::broadcast(comm_cart, distance, 0);
+std::vector<std::pair<int, int>>
+get_pairs_of_types(double const distance, std::vector<int> const &types) {
+  return get_pairs_filtered(distance, [types](Particle const &p) {
+    return std::any_of(types.begin(), types.end(),
+                       [p](int const type) { return p.p.type == type; });
+  });
+}
 
+void get_pairs_local(double const distance) {
   auto local_pairs = get_pairs(distance);
-
   Utils::Mpi::gather_buffer(local_pairs, comm_cart);
 }
 
-REGISTER_CALLBACK(mpi_get_pairs_local)
+REGISTER_CALLBACK(get_pairs_local)
 
-std::vector<std::pair<int, int>> mpi_get_pairs(double distance) {
-  mpi_call(mpi_get_pairs_local, 0, 0);
-  on_observable_calc();
+void get_pairs_of_types_local(double const distance,
+                              std::vector<int> const &types) {
+  auto local_pairs = get_pairs_of_types(distance, types);
+  Utils::Mpi::gather_buffer(local_pairs, comm_cart);
+}
 
-  boost::mpi::broadcast(comm_cart, distance, 0);
+REGISTER_CALLBACK(get_pairs_of_types_local)
+
+namespace detail {
+void search_distance_sanity_check(double const distance) {
```

Comment:
This can be implemented in a cell system agnostic way, by using ```ParticleDecomposition::max_range()```, also I think
this should be place inside the cell system code, not here, if there is no such check already.

Reply:
as far as I understand, ``max_range()`` refers to the local box length of the node without taking into consideration the cell decomposition within that local box

Reply:
Ok, you are right, not a good run for me so far :-). But then you this should be implemented, the way you are doing it encodes implementation details of the cell system at the call site. We were working hard for a long time to get those out, I think we shouldn't implement new ones. E.g. either each `ParticleDecomposition` should implement their own range check, or it should expose the maximum range that can be done without a resort.

Reply:
ok how about I rename ``max_range()`` to ``max_cutoff()`` to say which is the maximum cutoff that the chosen decomposition can handle; and then a new function ``max_range()`` to give the actual range that the calculations are performed in after the (local cell-) decomposition has been done?

Reply:
Ough, names are hard. But I'd use a new name for the new function, to not cause any unneeded confusion. I think `range` is better suited than `cutoff` (passive vs active), how about `current_range`?

Reply:
can do. But the doc for ``max_range()`` is ``@brief Maximum supported cutoff.`` :D

Reply:
Then I'd propose that you implemented `search_distance_sanity_check` in terms of this new function (without metioning the composition type), and we'll leave adding the check in a generic way to `CellStructure` for a later day.

Reply:
alright

Reply:
on a side note: Even for atom decomposition, the maximum cutoff is still bound by the global box size, right?

Reply:
I'd do `auto current_cutoff() { return max_cutoff(); }` :-), it's not actually bounded, because every pair is visited no matter the distance...